cirq/sim/density_matrix_simulator_test.py

```diff
@@ -1112,3 +1112,36 @@ def test_simulate_noise_with_terminal_measurements():
     result2 = simulator.run(circuit2, repetitions=10)
 
     assert result1 == result2
+
+
+def test_density_matrix_copy():
+    sim = cirq.DensityMatrixSimulator()
+
+    q = cirq.LineQubit(0)
+    circuit = cirq.Circuit(cirq.H(q), cirq.H(q))
+
+    matrices = []
+    for step in sim.simulate_moment_steps(circuit):
+        matrices.append(step.density_matrix(copy=True))
+    assert all(np.isclose(np.trace(x), 1.0) for x in matrices)
```

Comment:
Just testing the trace might not catch all non-copying conditions.  You could save each matrix and a copy then check they are all pairwise equal afterward.  Or you could check np.may_share_memory/np.shares_memory over all pairs.

Reply:
done